### DISCUSSION TITLE
FTorch Intel Compiler Flags

### DISCUSSION THREAD

**Discussion by backfelix8** [Cambridge-ICCS/FTorch#162] [Q&A] [answered]
Dear FTorch Community,

I am trying to set up FTorch on a rather large codebase on an HPC for a climate model (using the intel compilers).

Currently, when executing make, I receive the following error messages:
`ld: warning: libcuda.so.1, needed by <path>/FTorch_install/lib64/libftorch.so, not found (try using -rpath or -rpath-link)
lib/libicon.a(mo_art_linoz.o): In function mo_art_linoz_mp_art_calc_linoz_:
<path>/icon-kit/externals/art/chemistry/mo_art_linoz.f90:66: undefined reference to torch_model_load_
<path>/icon-kit/externals/art/chemistry/mo_art_linoz.f90:90: undefined reference to torch_model_forward_
<path>/icon-kit/externals/art/chemistry/mo_art_linoz.f90:93: undefined reference to torch_tensor_to_array_
<path>/icon-kit/externals/art/chemistry/mo_art_linoz.f90:103: undefined reference to torch_delete_
<path>/icon-kit/externals/art/chemistry/mo_art_linoz.f90:104: undefined reference to torch_delete_
<path>/icon-kit/externals/art/chemistry/mo_art_linoz.f90:107: undefined reference to torch_delete_
<path>/FTorch_install/lib64/libftorch.so: undefined reference to std::__throw_bad_array_new_length()@GLIBCXX_3.4.29
make[1]: *** [icon.mk:486: bin/icon] Error 1`

They indicate that the import of ftorch in my code is resolved, however, the actual functions are not being detected.

I am using the following compiler flags/directives:
`FTORCH_LIBPATH='/<path>/FTorch_install/lib64'`
`FTORCH_FCFLAGS='-I/<path>/FTorch_install/include/ftorch'`
`FTORCH_LIBS='-L/<path>/FTorch_install/lib64 -lftorch'`
`FTORCH_LDFLAGS='-L/<path>/FTorch_install/lib64'`

`BUILD_ENV="export LD_LIBRARY_PATH=\"${INTEL_ROOT}/lib:${INTEL_ROOT}/compiler/lib/intel64:${MPI_ROOT}/lib64:${CUDA_ROOT}/lib64:${HDF5_ROOT}/lib:${NETCDF_ROOT}/lib:${NETCDFF_ROOT}/lib:${ECCODES_ROOT}/lib64:${FYAML_ROOT}/lib:${MKL_ROOT}/lib/intel64:/usr/lib64:/usr/lib:${FTORCH_LIBPATH}\"; export PATH=\"${HDF5_ROOT}/bin:${NETCDF_ROOT}/bin/:/usr/bin\";"`
`LDFLAGS="-L${INTEL_ROOT}/lib -L${INTEL_ROOT}/compiler/lib/intel64 -L${MPI_ROOT}/lib64 -L${CUDA_ROOT}/lib64 -L${HDF5_ROOT}/lib -L${NETCDF_ROOT}/lib -L${NETCDFF_ROOT}/lib -L${MKL_ROOT}/lib/intel64 -L${ECCODES_ROOT}/lib64 -L${FYAML_ROOT}/lib ${FTORCH_LDFLAGS}"`
`LIBS="-Wl,--disable-new-dtags -Wl,--as-needed ${MPI_LIBS} ${XML2_LIBS} ${ZLIB_LIBS} ${ECCODES_LIBS} ${FYAML_LIBS} ${MKL_LIBS} ${NETCDFF_LIBS} ${NETCDF_LIBS} ${HDF5_LIBS} ${FTORCH_LIBS}"`
`FCFLAGS="-I${INTEL_ROOT}/include -I${MPI_ROOT}/include -I${MPI_FORTRAN_MOD_DIR} -I${HDF5_ROOT}/include -I${NETCDFF_ROOT}/include ${FTORCH_FCFLAGS} -I${ECCODES_ROOT}/include ${ARCH} -g -gdwarf-4 -pc64 -fp-model source"`

Do you have any suggestions on what I could try to resolve the error, so that all FTorch functions are imported correctly?

**Comment by jatkinson1000** [comment]
Hi @backfelix8 

A few ideas:
- Check that the ftorch install contains `lib64/` as you are using here - on some systems it is created as `lib/`.
- It also looks like there is a missing cuda library (though this is just a warning). If you need cuda can you import it, and if not you could try using the GPU-only version of libtorch.
- I would have perhaps expected an ` -lftorch` at the end of the `LDFLAGS`, though I'm not super familiar with the way things are set up on your particular system.

Let me know how that goes, and if no luck we can discuss further or perhaps arrange a call.

I can see that you are trying to couple to ICON, and we have collaborators who have successfully done this in the past, so we might be able to reach out to them.

**Reply by backfelix8** [reply]
Hi @jatkinson1000,

Thanks for the quick reply! I've (already) tried out the points, there is some more info that I can give you:
1. Yes, FTorch_install contains a lib64 folder with content
2. Yes, I would be able to include CUDA with an alteration of LD_LIBRARY_PATH, do you think this might fix the import issue or just the warning?
3. I have tried to set the -lftorch, as it is documented that way. However, when setting this option, the compiler throws an error... not completely sure what that means. In addition, all other entries in LDFLAGS also lack the -l* string

If you have any further ideas I would be very grateful

**Reply by jatkinson1000** [reply]
@backfelix8 

Yes - I think the way your machine/code sets up the `LDFLAGS` is different so perhaps the ` -lftorch` is not required there.

Further to my original comments it looks like the final error is being thrown from GLIBCXX.
Can you check that the version of `gcc` used when building ICON is the same version as was used when building FTorch? The error would suggest that there is a difference.
You can probably check this by setting up the ICON environment as above, and then comparing the `gcc` version there to the one in use when/where you build FTorch.

The fact this occurs at runtime not compile time suggests that there is a dynamic linking issue.
So first of all, can you run `ldd` ([man page here](https://www.man7.org/linux/man-pages/man1/ldd.1.html) if you need it) on your final executable to check that it is linked to FTorch and let us know the results?

**Reply by backfelix8** [reply]
@jatkinson1000 

Regarding the compilers:
I am actually building ICON with icc (and ifort) as the compilers. I also specified these compilers when configuring ftorch. On the HPC system I am using module to always use the 2023.1.0 version, so it should be the same version. (However, one thing to note is that when executing make --build... I have to load the gnu compilers on my HPC as the default gcc version is not recent enough. So if you think it might be an issue to execute make and make --build with different compilers loaded, I will investigate that further.

Regarding ldd:
I noticed that the formulation on my initial question was wrong. The error happens when executing make after executing the configuration script, so i do not have an executable yet. If you need any more data, such as the make output, just let me know.

**Reply by jatkinson1000** [reply]
Thanks @backfelix8 
So, the intel compilers will use gcc for the C++ templating library, so if you have a different gcc loaded when you build FTorch to the gcc loaded when you build icon this could lead to it being unable to find the right file to link. Can you try and ensure you use the same one for the entire process from start to finish?

The errors definitely look like a linking error, and that would make sense with your revised question. So my next guess would be to look at the `FTORCH_LIBS` and `FTORCH_LDFLAGS`. Have you got examples of what other environment variables used in `LIBS` and `LDFLAGS` look like? I am wondering if `FTORCH_LIBS` is the correct format given that the `-L` etc is specified outside the root environment variable elsewhere.

**Reply by backfelix8** [reply]
@jatkinson1000 
sure, i can give you an example, for MPI and FYAML:

`MPI_ROOT='/software/all/mpi/openmpi/5.0_intel_2023.1.0'`
`MPI_LIBS='-lmpi_usempif08 -lmpi_usempi_ignore_tkr -lmpi_mpifh -lmpi'`
`MPI_FORTRAN_MOD_DIR="${MPI_ROOT}/lib64"`

`FYAML_ROOT='/software/all/lib/libfyaml/0.9'`
`FYAML_LIBS='-lfyaml'`

`BUILD_ENV="export LD_LIBRARY_PATH=\"${INTEL_ROOT}/lib:${INTEL_ROOT}/compiler/lib/intel64:${MPI_ROOT}/lib64:${CUDA_ROOT}/lib64:${HDF5_ROOT}/lib:${NETCDF_ROOT}/lib:${NETCDFF_ROOT}/lib:${ECCODES_ROOT}/lib64:${FYAML_ROOT}/lib:${MKL_ROOT}/lib/intel64:/usr/lib64:/usr/lib:${FTORCH_LIBPATH}\"; export PATH=\"${HDF5_ROOT}/bin:${NETCDF_ROOT}/bin/:/usr/bin\";"`

`LIBS="-Wl,--disable-new-dtags -Wl,--as-needed ${MPI_LIBS} ${XML2_LIBS} ${ZLIB_LIBS} ${ECCODES_LIBS} ${FYAML_LIBS} ${MKL_LIBS} ${NETCDFF_LIBS} ${NETCDF_LIBS} ${HDF5_LIBS} ${FTORCH_LIBS}"`

I noticed that LIBS does not contain the paths to the software itself, only strings like `-lfyaml`. I tried changing that for ftorch by only adding `-lftorch` to LIBS instead of the path but that did not fix it. Maybe something else is wrong. Do you think that it makes a difference if e.g. the `-L` is within the FTorch variables or not?

As you mentioned that you know some collaborators that included FTorch into ICON specifically, it might be really helpful to take a look into their configuration file, though I of course don't know if they are available to you at the moment.

**Reply by jatkinson1000** [reply]
@helgehr are you able to advise how to set up a config file to use FTorch with ICON at all?
@backfelix8 is having issues linking during the build as described above.

**Reply by TomMelt** [reply]
@backfelix8 are you building this locally or on a HPC system? 

Would it be possible for you to send a zip of the source so that I could try building it?

**Reply by backfelix8** [reply]
Hi @TomMelt 

I am building ICON-ART on an HPC (HoreKa in Karlsruhe, Germany). Sure, I can try to give you the source code, but I am not completely confident that this will help, as ICON-ART has different configuration files for each venue and I do not know if the error can be reproduced on a different system. 

If you still want the code, could you perhaps provide an upload link (or any other preferred form), as the code base is quite large (5.8gb)

**Reply by helgehr** [reply, marked as answer]
Hello all, its been a while since I built this library but my setup for icon includes these two lines in the Makefile:
```
FCFLAGS += -I/path_to_ftorch/build/usr/local/include/ftorch
LDFLAGS += -L/path_to_ftorch/build/usr/local/lib64 -lftorch
```
Otherwise I can confirm the point from @jatkinson1000 that it is important to compile ftorch with the compilers (for fortran, c, and cpp) used in icon.

**Reply by backfelix8** [reply]
@jatkinson1000 @helgehr @TomMelt 

Thank you for your help, I was able to resolve the issue at least with the GNU compilers, so thank you again for mentioning the compilers to me.

The issue was a bit deeper than expected. I was using a torch version on my system that makes use of CUDA. So, while building FTorch with cmake, I was forced to use the GNU compilers in version 12, as version 13 is not compatible with CUDA (I am using module for software versioning, so it is not possible for me to use e.g. gfortran/13 and gcc/12 at the same time). However, the configuration script of ICON only allows GNU/13 to be used, which was causing the issue of non-detectable functions.

The solution was to install the CPU-only version of torch and re-building FTorch with GNU/13, which then allowed the project to compile.

**Reply by jatkinson1000** [reply]
Thanks @helgehr for your input, and thanks @backfelix8 for letting us know how you resolved it.

I hope you find FTorch useful, and please do get in touch if you need help, have feature requests, or have any scientific outputs you would like us to feature.
